core/os/device/deviceinfo/cc/query.h

```diff
@@ -25,6 +27,45 @@ namespace query {
 // current device. It must be freed with delete.
 device::Instance* getDeviceInstance(void* platform_data);
 
+// updateVulkanPhysicalDevices modifies the given device::Instance by adding
+// device::VulkanPhysicalDevice to the device::Instance. If a
+// vkGetInstanceProcAddress function is given, that function will be used to
+// resolve Vulkan calls, otherwise, all the Vulkan calls will be resolved from
+// Vulkan loader. If the given VkInstance handle is 0, a new VkInstance handle
+// will be created with the VkCreateInstance resolved from either the given
+// callback or the Vulkan loader.  The modified device::Instance will have its
+// ID re-hashed with the new content. Returns true if Vulkan physical device
+// info is fetched successfully and device::Instance updated, otherwise returns
+// false and keeps the device::Instance unchanged. Caution: When called with
+// VkGraphicsSpy layer loaded i.e. during tracing, the function pointer to a
+// layer under VkGraphicsSpy must be passed in. Resolving the Vulkan function
+// addresses from loader will cause a infinite calling stack and may deadlock
+// in the loader.
+bool updateVulkanDriver(
```

Comment:
Maybe change `getDeviceInstance` to take a struct of options and include vulkan info as part of that? GAPIS also calls this for host info, and it would be a nicer API if there was a single call that can return you the full proto message.

Reply:
Added an option arg to getDeviceInstance() in bc15aaa.
But I think I should keep this 'updateVulkanDriver()'. As when it is called in vkCreateInstance(), it probably will not be valid to get GL driver info again? Calling 'updateVulkanDriver' avoid touching other fields in the device::Instance

Reply:
This is almost what I was thinking, but not quite. We can work on this later.